Magnetic jamming detection

macaba opened the pull request.

Commit:
Added magnetic jamming detection & rejection

Commit:
Changed magnetic jamming algorithm to use a sliding window.

Commit:
Increased time constant of sliding window

Commit:
Conditional filter

If the magnetic magnitude is clearly wrong, don’t adjust the filter
value.

Commit:
Changed magMagnitudeFilteredMax

More real world testing needs to be done first, but in theory if the
average magMagnitudeFiltered were ever greater than 57, then the board
is in severe need of calibration.

Commit:
Further real world testing needed

sumotoy commented:
Well done! I love the magnetic jamming feature!

PaulStoffregen commented:
Long-term, I'd like to design this to use the magnetic field intensity from the calibration info.

For now, my main concern is the API.  I'd like to follow the "each thing added dilutes everything else" philosophy.  Is just the setMagneticJammingThreshold() function ok for now?  Or do we even need that?  If this stays, I'd really like to see it accept a percentage.  I can make some changes... but wanted to get your opinion first?

macaba commented:
> Long-term, I'd like to design this to use the magnetic field intensity from the calibration info.

Agreed although my testing shows that this might not actually be that useful hence the sliding window functionality. I'll calibrate in 1 room and get a field intensity of 42, and in another room I'll get a uniform field intensity of 48. This is definitely local man-made variation (room construction, steel lintels, large soft iron objects) but it's uniform over the whole room so I can still get a good heading reading.
I'll take a look at the calibration array source and see if I can figure out where to get the field intensity from but with the code as it is, it would instantly get overridden at Line 64 of MadgwickAHRS.cpp to cope with local area variation.

> I can make some changes... but wanted to get your opinion first?

Absolutely fine, this commit was really intended to get the ball rolling and go through a few cycles of code review. If you're happy with the general direction then I've got a few ideas to develop this further. I'd like to get your input as well and reduce your workload if you don't have to make any changes yourself.
- Make all the magnetic jamming code part of a MagneticJammingFilter class inside a MagneticJammingFilter.h file. Then there is only 1 additional variable inside the Madgwick class (of type MagneticJammingFilter) and any public methods could be accessed from the top level sketch like: 
  
  `filter.MagneticJamming.getJammingStatus();`
  
  This would also mean an easy inclusion into Mahony later with very little code duplication.
- I would like to get rid of the setMagneticJammingThreshold() by continually monitoring the RMS noise of the magnetic field intensity and dynamically setting the threshold to a fixed multiple of the noise. There's a possibility that this fixed multiple _could_ be a user settable value but testing may show this to be unnecessary.

What do you think of these ideas?

Edit: I accidentally clicked on the 'Close' button, sorry about that.

PaulStoffregen commented:
Yes, I like it.

Wouldn't a jamming filter class would need 2 public methods, one to get the status and another to update with a new intensity reading, right?

macaba committed:
Split out Magnetic Jamming functionality into seperate class

macaba committed:
Increased the range of the hard limits

It turns out that our South American friends have a very low magnetic
field intensity as seen on this map:

https://upload.wikimedia.org/wikipedia/commons/f/f6/World_Magnetic_Field_2015.pdf
I've had no reports of problems yet but I thought it would be good to
change the hard limits to correlate to actual geographical data, so the
variation of magnetic field intensity over major land masses (excluding
the poles) seems to be about 23uT to 62uT.

macaba commented:
Update:

I've split out the magnetic jamming filter which has made things very clean.

I'm still working on getting rid of the setMagneticJammingThreshold which also involves using the magnetic field intensity data from the EEPROM calibration data to set the _magMagnitudeFilteredMax_ and _magMagnitudeFilteredMin_ values which will make the jamming criteria more strict without becoming user-unfriendly.

macaba committed:
Now uses calibration data

Changed some terminology to match the MotionCal app.

macaba committed:
Added the ability to get the duration of the current state.

This allows the user to determine how long the IMU has been magnetically
jammed for, for warning/error purposes.
